src/mps/abstractprojmpo.jl

```diff
@@ -0,0 +1,242 @@
+      Hv *= P.H[j]
+    end
+    if !isnothing(rproj(P))
+      Hv *= rproj(P)
+    end
+  end
+  if order(Hv) != order(v)
+    error(
+      string(
+        "The order of the ProjMPO-ITensor product P*v is not equal to the order of the ITensor v, ",
+        "this is probably due to an index mismatch.\nCommon reasons for this error: \n",
+        "(1) You are trying to multiply the ProjMPO with the $(nsite(P))-site wave-function at the wrong position.\n",
+        "(2) `orthognalize!` was called, changing the MPS without updating the ProjMPO.\n\n",
+        "P*v inds: $(inds(Hv)) \n\n",
+        "v inds: $(inds(v))",
+      ),
+    )
+  end
+  return noprime(Hv)
+end
+
+(P::AbstractProjMPO)(v::ITensor) = product(P, v)
+
+"""
+    eltype(P::ProjMPO)
+
+Deduce the element type (such as Float64
+or ComplexF64) of the tensors in the ProjMPO
+`P`.
+"""
+function Base.eltype(P::AbstractProjMPO)
+  elT = eltype(P.H[P.lpos + 1])
+  for j in (P.lpos + 2):(P.rpos - 1)
+    elT = promote_type(elT, eltype(P.H[j]))
+  end
+  if !isnothing(lproj(P))
+    elT = promote_type(elT, eltype(lproj(P)))
+  end
+  if !isnothing(rproj(P))
+    elT = promote_type(elT, eltype(rproj(P)))
+  end
+  return elT
+end
+
+"""
+    size(P::ProjMPO)
+
+The size of a ProjMPO are its dimensions
+`(d,d)` when viewed as a matrix or linear operator
+acting on a space of dimension `d`. 
+
+For example, if a ProjMPO maps from a space with 
+indices `(a,s1,s2,b)` to the space `(a',s1',s2',b')` 
+then the size is `(d,d)` where 
+`d = dim(a)*dim(s1)*dim(s1)*dim(b)`
+"""
+function Base.size(P::AbstractProjMPO)::Tuple{Int,Int}
+  d = 1
+  if !isnothing(lproj(P))
+    for i in inds(lproj(P))
+      plev(i) > 0 && (d *= dim(i))
+    end
+  end
+  for j in (P.lpos + 1):(P.rpos - 1)
+    for i in inds(P.H[j])
+      plev(i) > 0 && (d *= dim(i))
+    end
+  end
+  if !isnothing(rproj(P))
+    for i in inds(rproj(P))
+      plev(i) > 0 && (d *= dim(i))
+    end
+  end
+  return (d, d)
+end
+
+function _makeL!(P::AbstractProjMPO, psi::MPS, k::Int)::Union{ITensor,Nothing}
+  # Save the last `L` that is made to help with caching
+  # for DiskProjMPO
+  L = nothing
+  while P.lpos < k
+    ll = P.lpos
+    if ll <= 0
+      L = psi[1] * P.H[1] * dag(prime(psi[1]))
+      P.LR[1] = L
+      P.lpos = 1
+    else
+      L = P.LR[ll] * psi[ll + 1] * P.H[ll + 1] * dag(prime(psi[ll + 1]))
+      P.LR[ll + 1] = L
+      P.lpos += 1
+    end
+  end
+  # Needed when moving lproj backward
+  P.lpos = k
+  return L
+end
+
+makeL!(P::AbstractProjMPO, psi::MPS, k::Int) = _makeL!(P, psi, k)
+
+function _makeR!(P::AbstractProjMPO, psi::MPS, k::Int)::Union{ITensor,Nothing}
+  # Save the last `R` that is made to help with caching
+  # for DiskProjMPO
+  R = nothing
+  N = length(P.H)
+  while P.rpos > k
+    rl = P.rpos
+    if rl >= N + 1
+      R = psi[N] * P.H[N] * dag(prime(psi[N]))
+      P.LR[N] = R
+      P.rpos = N
+    else
+      R = P.LR[rl] * psi[rl - 1] * P.H[rl - 1] * dag(prime(psi[rl - 1]))
```

Comment:
Possible to save a call to P.LR here, by either using R if that logic works out or else calling `rproj` ?

Reply:
Probably. I think using `rproj`, at least how it is design currently, would require `R` to be stored in the `Rcache` field. Probably there would be a way to make that work, but the goal was for this function to be generic between `DiskProjMPO` and `ProjMPO` (to minimize code duplication and make maintenance easier) and `ProjMPO` doesn't have an `Rcache` field.

Reply:
Though I guess `rproj` may automatically save the correct one in the cache? I have to draw out the diagrams.

Reply:
Alright, I improved caching in `makeL!` and `makeR!` by reusing `R`/`L` and also using `rproj`/`lproj`.